# Know your data

Learn how to effectively explore and deeply understand your data before applying datapreprocessing techniques on it.



---


> Author: Dr. Fadoua Ouamani


---





### **The Dataset**

The datas was collected via a form titled "Tell me more about you" that was shared with the student to collect :

*   Some details about their past grades in some classes related to datamining: Statistics, Mathematics, Programming, and English (as the class materials are all designed and written in english)
*   A personality test to have an overall idea about their personalities
*   Preferences

We will not be using all the data collected, we will only use the following data about:


*   The personlity type: nominal attribute with 4 values, Lion, Otter, Beaver, and Golden retriever
*   Being a collector or not: binary attribute
*   The age: Discrete numerical attribute
*   The gender: nominal with two values, F for female and M for male
*   Zodiac sign: nominal attribute with the values representing the different signs
*   Favorite language : nominal attribute as well and it may continue any language among the existing ones
*   4 past grades in Statistics, Mathematics, English and Object-Oriented Programming: Conitnous ration-scaled numerical attributes
*   Being someone who likes learning new skills or not: Binary attribute

The dataset is titled **TAYOU** as the abbreviation of Tell me About YOU. it was purely a matter of coincidence that this term has a beautiful meaning in Japanese. In fact, Tayou means diverse and varied as my students are <3




### 1.   Import the required libraries




In [ ]:
import pandas as pd    # to handle the dataframe
pd.set_option('expand_frame_repr', False) #to avoid the multi-lines formatting of the dataframe
import numpy as np     # to handle numbers and ndarray

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### 2.   Dataset loading, indexing and slicing



In [ ]:
# Charger le dataset
data_path = "/content/drive/MyDrive/big data/tayou.csv"
tayou_df = pd.read_csv(data_path)

In [ ]:
tayou_df.shape #...........................

(102, 11)

In [ ]:
tayou_df.columns #....................................

Index(['Personality', 'Collector', 'Age', 'ZodiacSign', 'FavLg', 'GStat',
       'GMath', 'GOOP', 'GEnglish', 'LearnNSkill', 'Gender'],
      dtype='object')

In [ ]:
tayou_df.head() #............................................

Personality Collector  Age ZodiacSign    FavLg  GStat  GMath  GOOP  GEnglish LearnNSkill Gender
0        Lion        No   24      Libra  English   15.0  15.00  15.0      18.0          No      M
1       Otter        No   23     Pisces   French    NaN    NaN   NaN       NaN          No      M
2        Lion        No   23        Leo  English   10.0  10.00  13.0      13.0         Yes      M
3        Lion        No   22    Scorpio  English   14.0  12.00  17.0      15.0         Yes      F
4        Lion        No   23        Leo   French   18.0  18.15  16.0      15.0         Yes      F

The dataset contains 102 rows and 11 columns. Below is a description of each column:

* **Personality:** Categorical data representing a personality type (e.g., "Lion",
"Otter").
* **Collector:** Categorical data indicating if the individual is a collector ("Yes" or "No").
* **Age**: Integer values representing the age of individuals.
* **ZodiacSign:** Categorical data for each person's zodiac sign (e.g., "Libra", "Pisces").
* **FavLg:** Categorical data representing the individual's favorite language (e.g., "English", "French"). This column has one missing value.
* **GStat:** Numerical data (float) representing a score for statistics. This column has some missing values.
* **GMath:** Numerical data (float) representing a score for mathematics. This column has some missing values.
* **GOOP:** Numerical data (float) representing a score for other subjects or topics. This column has some missing values.
* **GEnglish:** Numerical data (float) representing a score for English. This column has some missing values.
* **LearnNSkill:** Categorical data indicating if the individual is learning new skills ("Yes" or "No").
* **Gender:** Categorical data representing the individual's gender ("M" for male, "F" for female).

In [ ]:
#Indexing differently and slicing
print(tayou_df.loc[0].index) #.....................
tayou_df.set_index(np.arange(10,112), inplace=True) #...............................
print(tayou_df.loc[10].index) #..................................................
print('from 5 to 7 ',tayou_df.iloc[5:7,0:2]) # .................................................



Index(['Personality', 'Collector', 'Age', 'ZodiacSign', 'FavLg', 'GStat',
       'GMath', 'GOOP', 'GEnglish', 'LearnNSkill', 'Gender'],
      dtype='object')
Index(['Personality', 'Collector', 'Age', 'ZodiacSign', 'FavLg', 'GStat',
       'GMath', 'GOOP', 'GEnglish', 'LearnNSkill', 'Gender'],
      dtype='object')
from 5 to 7           Personality Collector
15              Lion       Yes
16  Golden retriever        No


In [ ]:
print(tayou_df.loc['15':'17', 'Personality':'Age']) #..................

         Personality Collector  Age
15              Lion       Yes   23
16  Golden retriever        No   23
17              Lion        No   24


We have changed the indices from 0-101 to 10-111.

 The .iloc[] method uses position-based indexing, meaning it ignores index labels and focuses solely on row positions. After setting a new index, the positions remain unchanged, so requesting rows at positions 5:7 will yield the 6th and 7th rows in the DataFrame, regardless of their index labels. In contrast, the .loc method performs label-based selection while .iloc performs position-based selection.

In [ ]:
tayou_df.head()

Personality Collector  Age ZodiacSign    FavLg  GStat  GMath  GOOP  GEnglish LearnNSkill Gender
10        Lion        No   24      Libra  English   15.0  15.00  15.0      18.0          No      M
11       Otter        No   23     Pisces   French    NaN    NaN   NaN       NaN          No      M
12        Lion        No   23        Leo  English   10.0  10.00  13.0      13.0         Yes      M
13        Lion        No   22    Scorpio  English   14.0  12.00  17.0      15.0         Yes      F
14        Lion        No   23        Leo   French   18.0  18.15  16.0      15.0         Yes      F

**Exercice 1:**

1.   Use the previous output to answer what is the difference in the behavior of .loc and .iloc when it comes to slicing the dataset.
2.   Without running but by only looking at the data, what will be the output of tayou_df.loc['10':'17', 'GStat':'GEnglish'].
3.   Without running but by only looking at the data, what will be the output of tayou_df.iloc[0:7, 5:8].

Answers:
1. he .iloc[] method uses position-based indexing, meaning it ignores index labels and focuses solely on row positions. In contrast, the .loc method performs label-based selection while .iloc performs position-based selection.
2. 15	15	15	18

  nan nan nan nan

   10	10	13	13

   14	12	17	15

   18	18.15	16	15

  16.5	12.2	13.62	14.5

  nan nan nan nan


3. English	15	15	15

  French nan nan nan

  English	10	10	13

  English	14	12	17

  French	18	18.15	16

  French	16.5	12.2	13.62

  English	15	16.7	16.5

In [ ]:
print(tayou_df.loc['10':'17','GStat':'GEnglish'])

    GStat  GMath   GOOP  GEnglish
10   15.0  15.00  15.00      18.0
11    NaN    NaN    NaN       NaN
12   10.0  10.00  13.00      13.0
13   14.0  12.00  17.00      15.0
14   18.0  18.15  16.00      15.0
15   16.5  12.20  13.62      14.5
16   15.0  16.70  16.50      13.5
17    NaN    NaN    NaN       NaN


In [ ]:
print(tayou_df.iloc[0:7,5:8])

    GStat  GMath   GOOP
10   15.0  15.00  15.00
11    NaN    NaN    NaN
12   10.0  10.00  13.00
13   14.0  12.00  17.00
14   18.0  18.15  16.00
15   16.5  12.20  13.62
16   15.0  16.70  16.50


In [ ]:
#Using loc, display only the rows with collector equal to 'Yes'
print(tayou_df.loc[tayou_df['Collector'] == 'Yes'])

          Personality Collector  Age   ZodiacSign    FavLg  GStat  GMath   GOOP  GEnglish LearnNSkill Gender
15               Lion       Yes   23       Taurus   French  16.50  12.20  13.62     14.50         Yes      M
19              Otter       Yes   23          Leo   Arabic  15.98  14.00  14.98     14.65         Yes      F
21               Lion       Yes   24      Scorpio  English  12.00  16.00  16.00     17.00         Yes      M
23              Otter       Yes   22  Sagittarius   Arabic  15.00  16.00  14.00     16.00         Yes      F
33               Lion       Yes   23        Aries  Spanish   8.40  11.00  17.26     14.18         Yes      F
41             Beaver       Yes   23        Libra   French  15.00  14.00  18.00     13.00         Yes      F
42   Golden retriever       Yes   24       Pisces  Italian  10.53  12.00  13.00     12.10         Yes      F
43               Lion       Yes   23       Taurus  Spanish  17.98  16.75  16.55     15.21         Yes      F
48             Beav

In [ ]:
#Using loc, display only the rows with collector equal to 'Yes' and LearnNSkill equal to 'Yes'
print(tayou_df.loc[(tayou_df['Collector'] == 'Yes') & (tayou_df['LearnNSkill'] == 'Yes')])

          Personality Collector  Age   ZodiacSign    FavLg  GStat  GMath   GOOP  GEnglish LearnNSkill Gender
15               Lion       Yes   23       Taurus   French  16.50  12.20  13.62     14.50         Yes      M
19              Otter       Yes   23          Leo   Arabic  15.98  14.00  14.98     14.65         Yes      F
21               Lion       Yes   24      Scorpio  English  12.00  16.00  16.00     17.00         Yes      M
23              Otter       Yes   22  Sagittarius   Arabic  15.00  16.00  14.00     16.00         Yes      F
33               Lion       Yes   23        Aries  Spanish   8.40  11.00  17.26     14.18         Yes      F
41             Beaver       Yes   23        Libra   French  15.00  14.00  18.00     13.00         Yes      F
42   Golden retriever       Yes   24       Pisces  Italian  10.53  12.00  13.00     12.10         Yes      F
43               Lion       Yes   23       Taurus  Spanish  17.98  16.75  16.55     15.21         Yes      F
48             Beav

In [ ]:
#interpret the results


* ** Personality Types:** The people listed come from various personality types such as Lion, Otter, Beaver, and Golden Retriever. This suggests that the interest in collection and learning new skills spans multiple personality categories.

* **Ages:** The ages of these individuals range from early 20s (22) to late 20s (29), which shows that people in this age group are more likely to be involved in both collection activities and learning new skills.

* **Zodiac Signs:** The individuals belong to diverse zodiac signs such as Taurus, Leo, Scorpio, etc. This indicates no clear correlation between zodiac signs and the tendencies toward being a collector and learning new skills.

* **Scores (GStat, GMath, GOOP, GEnglish):** Some individuals have missing scores in certain subjects (e.g., GStat, GOOP), but in general, they appear to have decent to high scores in most categories. This might suggest that collectors who are learning new skills may also perform well in their academic or personal achievements.

* **Gender:** The dataset includes both males and females in this group, showing no specific gender bias towards collecting and learning new skills.

In [ ]:
tayou_df.info
mlt_seris =tayou_df.groupby(['Personality','Gender','Collector']).GStat.mean()
mlt_seris

Personality       Gender  Collector
Beaver            F       No           13.553000
                          Yes          15.333333
                  M       No           13.000000
                          Yes          18.000000
Golden retriever  F       No           14.413571
                          Yes          11.643333
                  M       No           14.444444
                          Yes          17.000000
Lion              F       No           14.457692
                          Yes          13.076000
                  M       No           13.893125
                          Yes          13.828571
Otter             F       No           12.966667
                          Yes          14.326667
                  M       No            9.000000
                          Yes          13.330000
Name: GStat, dtype: float64

Note that you have created a **multi-index Series**, expand on its usefulness


Answer:

It provides the average GStat score for each unique combination of Personality, Gender, and Collector status. This allows us to see how different personality types, genders, and collector statuses perform on average in terms of their GStat score.

This can be useful for identifying patterns or trends across different groups, such as whether a particular personality or gender tends to have higher or lower average GStat scores among collectors or non-collectors.

**Exercice 2:**

Now run the following codes, study the outputs, and answer the questions.

1. When we use .iloc[] for a multi-index Series or DataFrame, what we should expect?
2. When we use .loc[] to access the data of one of the inner most index level of multi-index Series, what should we expect?
3. When we use .loc[] to access the data of one of non inner most index level of multi-index Series, what should we expect?
4. When you run either line of code below in the third cell, you will get an error and that is the point of this question. Study the error and try to answer the question.



In [ ]:
print(mlt_seris.iloc[0])
print(mlt_seris.iloc[1])
print(mlt_seris.iloc[2])

13.553
15.333333333333334
13.0


In [ ]:
mlt_seris.loc['Otter']

Gender  Collector
F       No           12.966667
        Yes          14.326667
M       No            9.000000
        Yes          13.330000
Name: GStat, dtype: float64

In [ ]:
print(mlt_seris.loc['F'])
print(mlt_seris.loc['Yes'])

KeyError: 'F'

Answers:
.................................................



1.   It will return the rows located at the specified position index, not based on the labels of the multi-index.

2.   It will return the values corresponding to that innermost level of the multi-index.

3.   It should return a sub-multi-index DataFrame or Series, still with the other levels of the multi-index intact.

4.   because the indexing method .loc[] was expecting a full hierarchical path through the multi-index levels.

### 3.   Central tendency



in this section we will be answering the following question: **Given an attribute, where do most of its values fall?**

The central tendency measures the location of **the middle **or **center **of a data distribution

Measures are: **mean**, **median** and **mode**


In [ ]:
num_split=tayou_df.loc[:,tayou_df.dtypes!=object] #..........................................
num_split.head()

In [ ]:
categ_split=tayou_df.loc[:,tayou_df.dtypes==object] #...........................................
categ_split.head()

In [ ]:
for column_name in num_split.columns:
    print(num_split[column_name].mean())
    print(num_split[column_name].median())
    print(num_split[column_name].mode())

 Interpret the obtained results: (Indication: compare the values you have obtained for each attribute for the mean, the median and the mode)

==> The mean and median are relatively close for all the numerical columns, indicating that the data is fairly symmetric.
The mode shows that the value of 15 is the most common in all columns, suggesting that a score of 15 is frequently repeated across different subjects.

In [ ]:
import matplotlib.pyplot as plt
import scipy.stats as stats
# Calculating the mean and standard deviation of the parameter "GStat":
mean = tayou_df['GStat'].mean()
std = tayou_df['GStat'].std()

# Calculating probability density function (PDF)
pdf = stats.norm.pdf(tayou_df['GStat'].sort_values(), mean, std)

# Drawing a graph
plt.plot(tayou_df['GStat'].sort_values(), pdf)
plt.xlim([5,20])
plt.xlabel('Grades in Statistics (GStat)', size=12)
plt.ylabel('Frequency', size=12)
plt.grid(True, alpha=0.3, linestyle='--')
plt.show()

Interpret the resulting plot:

The plot shows a bell-shaped curve, indicating that the grades in statistics (GStat) follow a normal distribution. Most students scored around the mean of 14, with fewer students scoring much higher or lower. The distribution is symmetrical, suggesting no significant skewness or outliers in the data.

In [ ]:
for column_name in categ_split.columns:
    print(categ_split[column_name].mode())
    print(categ_split[column_name].value_counts())

Interpret the results:
*  Lion is the dominant personality, and most respondents are not collectors.
*  Gemini and Leo are the most frequent zodiac signs.
*  English is the most preferred language, with some data inconsistencies in language entries.
*  The overwhelming majority enjoy learning new skills, and there is a slightly higher number of females than males.

### 4.  Dispersion



In this section, we will be answering the following question: **How are the data spread out?** by measuring the variability of data

Measures: **range**, **quartiles**,**interquartile range**, **variance** and **standard deviation**

In [ ]:
# the range
def range_imp(dat):
  return (dat.max()-dat.min())

In [ ]:
#Use the function to compute the range for each numerical attribute
numerical_columns = ['GStat', 'GMath', 'GOOP', 'GEnglish']
range_values = tayou_df[numerical_columns].apply(range_imp)

range_values

In [ ]:
tayou_df.describe()

In [ ]:
#Quartiles
for column_name in num_split.columns:
    print(num_split[column_name].quantile([0.25,0.5,0.75]))

In [ ]:
#interquartile range (IQR)
def interquartile_range(dat):
  return (dat.quantile(0.75)-dat.quantile(0.25))

In [ ]:
#Use the function to compute the interquartile range for each numerical attribute
iqr_values = tayou_df[numerical_columns].apply(interquartile_range)

iqr_values

In [ ]:
#boxplots
num_split.boxplot()

Interpret the boxplots accroding to the values obtained in the previous results

*  The grades generally hover around the median score of 14-15 across most subjects.
*  Statistics (GStat) has the most variability, with a wider range and more extreme low-performing students.
*  Mathematics (GMath) and English (GEnglish) show the presence of outliers, though the majority of the students perform within a close range.
*  There are several outliers across the dataset, indicating students who scored significantly higher or lower than the rest.

In [ ]:
#Variance
num_split.var()

In [ ]:
#Standard deviation
num_split.std()

Interpret the results:

GStat (Statistics) has the highest spread, indicating the most diverse performance among students.

GOOP (Programming) has the least variability, indicating more consistent performance.

The standard deviations for all subjects suggest that students' grades generally fall within a range of 2 to 3 points around the mean, but with some subjects showing more variation than others.

### 5. Symmetry

In [ ]:
#Skewness
num_split.skew()

Interpret the results:

Age is highly positively skewed, indicating that most students are younger, with some significantly older outliers.

The grades for GStat, GMath, GOOP, and GEnglish all show slight to moderate negative skewness, suggesting that more students scored on the higher end, but a few lower scores pull the tail to the left.


In [ ]:
#Kurtosis
num_split.kurtosis()

Interpret the results:

Age has an extremely leptokurtic distribution, with many outliers (students older than the majority).

GStat and GOOP have flatter (platykurtic) distributions, meaning fewer extreme values.

GMath and GEnglish show slight leptokurtosis, indicating more outliers or extreme values compared to a normal distribution.
